### PR TITLE
NOTMAD factor graphs

### PR DESCRIPTION
**Pull request by blengerich**
To scale NOTMAD, it will be helpful to integrate factor graphs. This implementations creates a new attribute `factor_mat_raw` on initialization of the `NOTMAD` lightning module. The number of factors (latent dimensions) is governed by the parameter `num_factors` (values <= 0 or = X.shape[-1] disable factor graphs.

Presently, the tests verify that factor graphs train and converge to a solution that is better than initialization, but extensive testing and simulations are required to see how much they help in dimensionality reduction.

### CODE REVIEW

**Review by cnellington — COMMENTED** [review]
This all looks great, I think there's two steps before merging:
1. We should verify that factor graphs work as expected. I believe we've already run simulations to check that the low-dim factor approx does worse than full-dim when true_dims > factor_dims. We should also check that factor graphs do better than full dim graphs when the true graphs only have a few factors (true_dims == factor_dims). 
2. We should give a factors kwarg to the easy Bayesian networks.

I'm pulling the SKLearnWrapper changes into dev/main separately since the predict_params bug is causing problems.

File: contextualized/dags/lightning_modules.py [hunk 1]
```diff
@@ -289,23 +337,33 @@ def training_epoch_end(self, training_step_outputs, logs=None):
...
+            and dag_params["alpha"] < 1e12
+            and dag_params["rho"] < 1e12
         ):
-            self.ss_alpha = self.ss_alpha + self.ss_rho * epoch_dag_loss
-            self.ss_rho = self.ss_rho * 1.1
-        self.h_old = epoch_dag_loss
+            dag_params["alpha"] = (
+                dag_params["alpha"] + dag_params["rho"] * epoch_dag_loss
+            )
+            dag_params["rho"] = dag_params["rho"] * 1.1
```
Comment (cnellington) [inline contextualized/dags/lightning_modules.py:364]: I think we might want to add the update "rate" (1.1 by default) as a loss parameter. This seems to have interplay with early stopping parameter choices.

File: contextualized/dags/tests.py [hunk 2]
```diff
@@ -157,24 +85,51 @@ def _evaluate(self, train_preds, test_preds, val_preds):
...
+        INIT_MAT = np.random.uniform(-0.1, 0.1, size=(k, 4, 4)) * np.tile(
+            1 - np.eye(4), (k, 1, 1)
+        )
         model = NOTMAD(
             self.C.shape[-1],
             self.X.shape[-1],
-            init_mat=INIT_MAT,
-            num_archetypes=model_args.get("num_archetypes", k),
-            num_factors=model_args.get("num_factors", 0),
+            archetype_params={
```
Comment (cnellington) [inline contextualized/dags/tests.py:94]: Might want to import the default args for our tests

File: contextualized/dags/lightning_modules.py [hunk 3]
```diff
@@ -22,21 +74,17 @@ def __init__(
...
-        use_dynamic_alpha_rho=True,
-        sample_specific_loss_params={"l1": 0.0, "alpha": 1e-1, "rho": 1e-2},
-        archetype_loss_params={"l1": 0.0, "alpha": 0.0, "rho": 0.0},
-        learning_rate=1e-3,
-        opt_step=50,
-        encoder_type="mlp",
-        encoder_kwargs={"width": 32, "layers": 2, "link_fn": identity_link},
-        init_mat=None,
+        sample_specific_params=DEFAULT_SS_PARAMS,
+        archetype_params=DEFAULT_ARCH_PARAMS,
```
Comment (cnellington) [inline contextualized/dags/lightning_modules.py:78]: I prefer `sample_specific_loss_params` and `archetype_loss_params` since this makes it seem like separate args like `num_archetypes` would be contained in `archetype_params`.

**Review by cnellington — COMMENTED** [review]
Looks amazing! Notebook docs look great. Very minor comments. The convergence tests are passing and I think this should be ready to merge. We'll do more extensive simulation studies offline.

File: contextualized/dags/lightning_modules.py [hunk 4]
```diff
@@ -188,24 +278,32 @@ def test_step(self, batch, batch_idx):
...
+            X_pred = dag_pred_with_factors(x_true, w_pred, self._factor_mat())
+        else:
+            X_pred = dag_pred(x_true, w_pred)
         mse_term = 0.5 * x_true.shape[-1] * mse_loss(x_true, X_pred)
         l1_term = l1_loss(w_pred, self.ss_l1).mean()
         # ignore archetype loss, use constant alpha/rho upper bound for validation
-        dag_term = dag_loss(w_pred, 1e12, 1e12).mean()
-        loss = mse_term + l1_term + dag_term
+        dag_term = self.ss_dag_loss(w_pred, **self.val_dag_loss_params).mean()
+        factor_mat_term = actor_mat_term = l1_loss(
```
Comment (blengerich) [inline contextualized/dags/lightning_modules.py:297]: Fix this typo

**Comment by blengerich** [comment]
TODO: Check if P^T should actually be a separate matrix Q.

**Comment by blengerich** [comment]
> This all looks great, I think there's two steps before merging:
> 
> 1. We should verify that factor graphs work as expected. I believe we've already run simulations to check that the low-dim factor approx does worse than full-dim when true_dims > factor_dims. We should also check that factor graphs do better than full dim graphs when the true graphs only have a few factors (true_dims == factor_dims).
> 2. We should give a factors kwarg to the easy Bayesian networks.
> 

There is now a demo notebook `easy_bayesian_network_factors.ipynb` which has a very basic investigation of the factor MSEs.

`num_factors` kwarg is now implemented for the easy Bayesian networks.

File: contextualized/dags/lightning_modules.py (hunk 1 above)
Comment (blengerich) [reply contextualized/dags/lightning_modules.py]: Resolved

File: docs/demos/easy_classification.ipynb [hunk 5]
```diff
@@ -69,3579 +69,9 @@
    "cell_type": "code",
```
Comment (cnellington) [inline docs/demos/easy_classification.ipynb:69]: <pre class="ql-syntax" spellcheck="false">Line #11.    Y = np.round(1. / (1+np.exp(-Y)))
</pre><p>Small issue with the simulation, this sets Y to the most likely class given X &amp; beta, but I believe we should be sampling Y from P(Y = 1) = 1 / (1 + np.exp(-Y_{line_10})). Check out the relationship between true and pred betas below, I think the direct assignment of Y to the most likely class is causing some major deviation between the true and pred.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/cnellington/Contextualized/pull/143/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (cnellington) [reply docs/demos/easy_classification.ipynb:69]: Looks like this is already online and not specific to this PR. I've made a separate issue #169 to address this.

Comment (blengerich) [reply docs/demos/easy_classification.ipynb:69]: PR filed to fix this: #170
